From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: bsp <bsp@radxa.com>
Date: Fri, 16 Jun 2023 19:56:14 +0800
Subject: [PATCH] Enable 1.8V SD data rates

This is currently disabled since U-Boot does not support 1.8V signaling, thus
failed to read the card after reboot.

---
 .../dts/amlogic/meson-g12a-radxa-zero.dts     | 22 +++++++++++++++++--
 .../dts/amlogic/meson-g12b-radxa-zero2.dts    | 22 +++++++++++++++++--
 2 files changed, 40 insertions(+), 4 deletions(-)

diff --git a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts b/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts
index 8f0fbaa1dc826..bc9d8c5c6ce12 100644
--- a/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-g12a-radxa-zero.dts
@@ -152,6 +152,19 @@ vddao_3v3: regulator-vddao_3v3 {
 		regulator-always-on;
 	};
 
+	vddio_c: regulator-vddio-c {
+		compatible = "regulator-gpio";
+		regulator-name = "VDDIO_C";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+
+		gpios = <&gpio_ao GPIOAO_7 GPIO_ACTIVE_HIGH>;
+		gpios-states = <0>;
+		states = <3300000 0>,
+				 <1800000 1>;
+	};
+
 	vddcpu: regulator-vddcpu {
 		compatible = "pwm-regulator";
 
@@ -441,12 +454,17 @@ &sd_emmc_b {
 
 	bus-width = <4>;
 	cap-sd-highspeed;
-	max-frequency = <100000000>;
+	sd-uhs-sdr12;
+	sd-uhs-sdr25;
+	sd-uhs-sdr50;
+	sd-uhs-ddr50;
+	sd-uhs-sdr104;
+	max-frequency = <200000000>;
 	disable-wp;
 
 	cd-gpios = <&gpio GPIOC_6 GPIO_ACTIVE_LOW>;
 	vmmc-supply = <&vddao_3v3>;
-	vqmmc-supply = <&vddao_3v3>;
+	vqmmc-supply = <&vddio_c>;
 };
 
 /* eMMC */
diff --git a/arch/arm64/boot/dts/amlogic/meson-g12b-radxa-zero2.dts b/arch/arm64/boot/dts/amlogic/meson-g12b-radxa-zero2.dts
index 567a443356fe3..bf5d43d48f22a 100644
--- a/arch/arm64/boot/dts/amlogic/meson-g12b-radxa-zero2.dts
+++ b/arch/arm64/boot/dts/amlogic/meson-g12b-radxa-zero2.dts
@@ -131,6 +131,19 @@ vddao_3v3: regulator-vddao-3v3 {
 		regulator-always-on;
 	};
 
+	vddio_c: regulator-vddio-c {
+		compatible = "regulator-gpio";
+		regulator-name = "VDDIO_C";
+		regulator-min-microvolt = <1800000>;
+		regulator-max-microvolt = <3300000>;
+		regulator-always-on;
+
+		gpios = <&gpio_ao GPIOAO_2 GPIO_ACTIVE_HIGH>;
+		gpios-states = <0>;
+		states = <3300000 0>,
+				 <1800000 1>;
+	};
+
 	vddcpu_a: regulator-vddcpu-a {
 		/*
 		 * MP8756GD Regulator.
@@ -451,12 +464,17 @@ &sd_emmc_b {
 
 	bus-width = <4>;
 	cap-sd-highspeed;
-	max-frequency = <50000000>;
+	sd-uhs-sdr12;
+	sd-uhs-sdr25;
+	sd-uhs-sdr50;
+	sd-uhs-ddr50;
+	sd-uhs-sdr104;
+	max-frequency = <200000000>;
 	disable-wp;
 
 	cd-gpios = <&gpio GPIOC_6 GPIO_ACTIVE_LOW>;
 	vmmc-supply = <&vddao_3v3>;
-	vqmmc-supply = <&vddao_3v3>;
+	vqmmc-supply = <&vddio_c>;
 };
 
 /* eMMC */
-- 
2.41.0

